### PR TITLE
Clown Car balance changes

### PR DESCRIPTION
**Pull request by lilacrhododendron**
:cl:
balance: The Clown Car can no longer move unrestricted in zero gravity environments.
balance: The Clown Car now costs 20 TC to purchase.
/:cl:

This PR was done at the request of @vuonojenmustaturska. I renamed entered.dm to sealed.dm because it's where the sealed subtype is defined and it's primarily used for procs specific to sealed. The balance changes were Naksu's idea. I definitely think not being able to move freely in space is a good change. I'm ambivalent about the TC, but the item felt like a gimmick that you should fully commit to like His Grace, and the free space movement + 4 TC space suit was a bit cancerous.

### CODE REVIEW

**Comment by obscolene** [comment]
\>"""minor balance changes"""
\>just a heavy nerf with no upsides

big laffo

**Comment by ExcessiveUseOfCobblestone** [comment]
@Qustinnus

**Comment by subject217** [comment]
It's still a pretty good gimmick item, as it should be. I mostly just made it because Naksu was complaining about it and I could see his point. I'm sure Qustinnus does not want his precious car nerfed.

**Comment by MrDoomBringer** [comment]
removing interesting traitor synergies is a good thing now? Maybe reconsider the change if your only rationale is "you should fully commit to it".

 Should we expect people to fully commit to owning a desword, or fully commit to using sleeping carp, when these items (just like the car) have already been balanced around limiting the possibilities of other gear that you can buy in conjunction?

**Comment by Qustinnus** [comment]
Changing tc price is a dumb change, the other thing is a bugfix of kevinz code at best

**Comment by Bawhoppen** [comment]
Changing the TC price of something isn't often the best way to balance something.

**Comment by kriskog** [comment]
How are nukies going to get their clown cars to the station now?

**Comment by Mickyan** [comment]
The TC increase is uncalled for and the reasoning for it makes absolutely no sense

**Comment by subject217** [comment]
Justify it being 15 TC, then. As far as I'm concerned it's more than powerful enough to be 20 TC.

**Comment by GuyonBroadway** [comment]
Its nowhere as strong or lethal as items that cost less than 20tc.

See dual-e and Sleeping Carp for examples.

**Comment by Qustinnus** [comment]
you already nerfed it with the spacewalk thing. Your only reasoning for making it 20 was that you could buy the spacesuit but that's no longer an issue.

Also we don't have to explain why 15 is better because we're not the one trying to change the cost

**Comment by Ziiro** [comment]
Make it so using a jetpack on it gives it spaceflight

**Comment by Qustinnus** [comment]
@Dorsisdwarf what would you turn the glomp sounds into because idk if i trust you

**Comment by Suicidalpickles** [comment]
I personally find the clown car amusing but do think its a bit too cheap tbh for how it performs, wouldn't mind an up in price.

**Comment by subject217** [comment]
>Also we don't have to explain why 15 is better because we're not the one trying to change the cost

Yeah, except you did change the cost. In a separate PR, when you applied a totally fair health nerf that could have stood on its' own, and in the midst of other buffs too like making the car be repair-able with bananas. Care to explain the rationale behind that? Because to me it looks like the reason that I suggested, that you are trying to stealth-buff it by making really overpowered shit locked behind the emag, and then making it considerably easier to acquire an emag for it. You have another emagged clown car buff PR up at this very moment, in fact.

It is a no slowdown vehicle that grants total immunity to all stuns and has 150 repairable HP. It also lets you fucking car vore people by running them over. If you try to shoot at it instead it will shit lube, good luck keeping up while also shooting and not slipping. If you try to use a melee weapon you are within VORING DISTANCE and do not have long to go before you are swallowed up. Not to mention that it only drops you out when you collide into walls, and that it will ONLY work for the clown, and nobody else. With an emag, the item just gets even more ridiculous. It is pretty much the same as His Grace (which isn't actually a good item, because it always ends up biting you in the ass, literally) in terms of purpose. Epic meme items shouldn't be the meta choice, that makes them stale. The rareness and the absurdity of the gimmick is what makes it interesting.

**Comment by epochayur** [comment]
tl;dr

**Comment by MrDoomBringer** [comment]
image

**Comment by subject217** [comment]
@MrDoomBringer I've explained my points very clearly and explicitly, and apparently I am now posting too much because I am bothering to think things through rationally. Meanwhile you have a downvote button. Very impressive.

**Comment by Ziiro** [comment]
So the point is that while it's a meme item at 15 TC and very fucking annoying to deal with, it's hard for it to be lethal unless the clown drives you into space. Yes, fixing that is cool. Do that.

Increasing the cost **and** decreasing the lethality at the same time is what makes this a bad PR.

**Comment by subject217** [comment]
That part is actually going to probably require a refactor to work properly, I have to look at riding datums cause vehicles are shit code. Being 20 TC doesn't make it less of a meme item and it doesn't make it less annoying to deal with. What it means is that you can't get misc traitor gear with the clown car and it's significantly harder to acquire an emag with the clown car. It also makes traitors less likely to give the clown their PDAs for le epic memay if it means they lose all their TC. The price buff to 15 TC was completely uncalled for and the universally negative reaction to the price nerf shows how little thought anyone else has put into this. A price nerf is one of the most gentle and rational nerfs to the car. The banana healing could easily have an action bar tied to it, or some sort of cap for it. The car working for only clowns is extremely overpowered and reminiscent of CM skill locks. The VOREMOBILE 9000 is far better than His Grace, and I would say it's easily equal to Sleeping Carp or Desword + noslips if you know how to use it.

**Comment by epochayur** [comment]
>The price buff to 15 TC was completely uncalled for and the universally negative reaction to the price nerf shows how little thought anyone else has put into this.

lol
just because everyone else doesn't argue in 2000-word essays doesn't mean they haven't thought about it

**Comment by subject217** [comment]
The meaningful difference is reducing the probability of emagged clown cars. It's pretty clear dude.

**Comment by Qustinnus** [comment]
lol are you fwoosh in disguise

anyways thanks for showing how mature you are by going from balance talk to shittalking my code and changelogs, anything i didn't put in my changelog i forgot to add, stop pretending like im some kind of stealthy madman thats trying to sneak in the evil 15tc change, i'm fairly sure i can change my feature based on feedback. you cant even buy an emag with a 15tc price

**Comment by Qustinnus** [comment]
`That part is actually going to probably require a refactor to work properly, `
lol you shit talk my coding but you think that requires a refactor, i could add it to my pr and you can close his pr

**Comment by subject217** [comment]
You're not the one who wrote the code that I'm mad at, at any rate. My terminology is fucked but if you mean using a riding component it is in fact very simple and easy to do, as you say, which I found out the moment that I actually looked at how ridden works.  Seems like a band-aid to me though. Also, if you do add it to your PR you should rename entered.dm :^)

File: code/modules/vehicles/cars/car.dm [hunk 1]
```diff
@@ -1,18 +1,34 @@
...
 	var/car_traits = NONE //Bitflag for special behavior such as kidnapping
 	var/engine_sound = 'sound/vehicles/carrev.ogg'
 	var/last_enginesound_time
 	var/engine_sound_length = 20 //Set this to the length of the engine sound
 	var/escape_time = 200 //Time it takes to break out of the car
 
+/obj/vehicle/sealed/car/Initialize()
+	. = ..()
+	var/datum/component/riding/D = LoadComponent(/datum/component/riding)
+	D.vehicle_move_delay = 2
```
[the remark this answers is not in the document]
Comment (subject217) [reply code/modules/vehicles/cars/car.dm]: It's the tick delay between movements. Don't see what you mean by magic numbers. I actually don't even need this here at all though because for some reason I thought the default in the component was 1. I'll just get rid of it :^)

File: code/modules/vehicles/cars/car.dm [hunk 2]
```diff
@@ -1,18 +1,34 @@
...
 /obj/vehicle/sealed/car/generate_actions()
 	. = ..()
 	initialize_controller_action_type(/datum/action/vehicle/sealed/remove_key, VEHICLE_CONTROL_DRIVE)
 	if(car_traits & CAN_KIDNAP)
 		initialize_controller_action_type(/datum/action/vehicle/sealed/DumpKidnappedMobs, VEHICLE_CONTROL_DRIVE)
 
+/obj/vehicle/sealed/car/driver_move(mob/user, direction)
+	if(key_type && !is_key(inserted_key))
+		to_chat(user, "<span class='warning'>[src] has no key inserted!</span>")
+		return FALSE
```
[the remark this answers is not in the document]
Comment (subject217) [reply code/modules/vehicles/cars/car.dm]: It's how `/obj/vehicle/ridden` does it even though the key functionality is inherited from `vehicle`. I figured it would function basically the same here.
From vehicle:
 ```
/obj/vehicle/proc/driver_move(mob/user, direction)
	if(key_type && !is_key(inserted_key))
		to_chat(user, "<span class='warning'>[src] has no key inserted!</span>")
		return FALSE
	if(!default_driver_move)
		return
vehicle_move(direction)
```

[the remark this answers is not in the document]
Comment (Qustinnus) [reply code/modules/vehicles/cars/car.dm]: thats kevinz being a dumb

**Comment by MrDoomBringer** [comment]
@subject217 hey man i explained my thoughts clearly and explicitly in my first response as well theres no need to be sarcastic

File: code/modules/vehicles/cars/car.dm (hunk 1 above)
[the remark this answers is not in the document]
Comment (Qustinnus) [reply code/modules/vehicles/cars/car.dm]: magic numbers as in the 2 is a random number thats not being pulled from anywhere, for example I imagine we want the vehicle_move_delay to be our vehicles move delay, instead of 2.

[the remark this answers is not in the document]
Comment (subject217) [reply code/modules/vehicles/cars/car.dm]: Oh yeah, I understand what you mean. I'll change it.

**Comment by Iamgoofball** [comment]
>balance: The Clown Car now costs 20 TC to purchase.

this removes the emag point you idiot

make it cost 20 if you bought it with an emag

**Comment by subject217** [comment]
>this removes the emag point you idiot

yep the point is you can't get it unless you loot one off another traitor somehow or convince one to emag it for you. if it was easy to attain everyone would just always get the emag with it, the emag is an incredibly useful item in its own right anyway.

**Comment by MrDoomBringer** [comment]
ok ive carefully deduced that subject has a fair point but the meme image is staying

File: code/modules/vehicles/cars/car.dm (hunk 2 above)
[the remark this answers is not in the document]
Comment (ninjanomnom) [reply code/modules/vehicles/cars/car.dm]: @Qustinnus is right. This related code should be changed to properly be able to support additional move restrictions without duplicating code.

File: code/modules/uplink/uplink_items.dm [hunk 3]
```diff
@@ -1480,7 +1480,7 @@ GLOBAL_LIST_INIT(uplink_items, subtypesof(/datum/uplink_item))
 			someone saves them or they manage to crawl out. Be sure not to ram into any walls or vending machines, as the springloaded seats \
 			are very sensetive. Now with our included lube defense mechanism which will protect you against any angry shitcurity!"
 	item = /obj/vehicle/sealed/car/clowncar
-	cost = 15
+	cost = 20
```
Comment (ninjanomnom) [inline code/modules/uplink/uplink_items.dm]: Personally I find this unnecessary but feel free to pr this by itself. Take it out of this pr which is functionally a bug fix.

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by Qustinnus** [comment]
did you actually test this

**Comment by subject217** [comment]
Yes, it functions.

**Comment by epochayur** [comment]
dumb

**Comment by Ziiro** [comment]
`>It's another episode of "merged despite overwhelming opinion to the contrary"`

**Comment by Qustinnus** [comment]
@optimumtact how is a tc value of 20 something that should have been in the original pr lol

**Comment by subject217** [comment]
Based and redpilled.

**Comment by Qustinnus** [comment]
and the PR wasn't even done the code is actually worse now so now I can either 1. revert this or 2. make a fix pr

**Comment by subject217** [comment]
make a fix, easy GBP courtesy of me

**Comment by Qustinnus** [comment]
https://streamable.com/hqcwz

**Comment by Qustinnus** [comment]
im glad you lied when i explicitly asked if you tested this PR though

**Comment by optimumtact** [comment]
@Ziiro I know you're from oracle and probably confused about how things work here, but we don't just take into account popular opinion

File: code/modules/vehicles/cars/car.dm (hunk 1 above)
[the remark this answers is not in the document]
Comment (Qustinnus) [reply code/modules/vehicles/cars/car.dm]: you clearly didnt

**Comment by subject217** [comment]
I tested it right before I pushed my latest commit, and it worked as intended. Then I was changing the *magic numbers* stuff and pushed a commit. I was going to work on it more today probably but then it got merged, so.

**Comment by Qustinnus** [comment]
it didn't work as intended before that either because you set the movedelay to 2. stop bullshitting.

**Comment by subject217** [comment]
Was that not what it was always at? If so, it was an honest mistake. Please fix it, the original movespeed was fine.

**Comment by Qustinnus** [comment]
no 2 is the default value so by setting it to 2 you didn't actually change anything. i explained to you precisely what to change. im not getting out of bed now to make a fix pr so someone else can go do it.

**Comment by subject217** [comment]
I was always under the impression that the tick move delay of the car was 2 because that's the vehicle default and I didn't see the var changed anywhere along the path. If I was mistaken, then I'll fix it.

**Comment by Qustinnus** [comment]
attached you will find a picture of the movedelay, i added some arrows so you wont miss it this time. leaving aside the fact that even if none of the vehicles on path changed the movedelay, it would still be shit because by adding the riding component you /still/ had to change the move delay to match the vehicle move delay.

image

**Comment by subject217** [comment]
Nice, I'm just blind and I misunderstood you before. I will PR a fix real fast.

**Comment by adamkad1** [comment]
and now that we got crawling added they should probally remove this... though im sure oranges the fun hater disagrees

**Comment by MrDoomBringer** [comment]
image

**Comment by MrDoomBringer** [comment]
image

**Comment by Qustinnus** [comment]
is this my daily reminder that I still dislike subject
